### PR TITLE
Jigsaw Surface

### PR DESCRIPTION
**Pull request by busterbeam**
Sorry @EarthKiii I think I may of slightly broken the edge snapping when refactoring the code.

May I suggest you look into `pygame`'s Vector object it allows arithmetic's with tuples much more easy.

I've spread it all across 3 files still `__init__.py` there but `jigsaw.py` has the `JigSaw` object separate and `tiles.py` has to objects `Tile` and `JigSawTiles`.  I hope this makes more sense or at least more readable. separated ~300 lines of code into 3, ~100 line files

### CODE REVIEW

File: jigsaw-game/common/tessellation.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+    array,
+    column_stack,
+    hstack,
+    meshgrid,
+    ones,
+    random,
+    sqrt,
+    vstack,
+    zeros,
+)
```
Comment (LukeHankey) [inline jigsaw-game/common/tessellation.py:17]: I think this would be better as just `import numpy as np`. I don't think I have ever seen importing so many functions from numpy before. 👀

File: requirements.txt [hunk 2]
```diff
@@ -2,3 +2,4 @@
 pillow~=10.0.0
 pygame~=2.5.1
 pre-commit~=3.4.0
+scipy~=1.11.2
```
Comment (LukeHankey) [inline requirements.txt:5]: Will also want to add numpy as well.

**Review by LukeHankey — CHANGES_REQUESTED** [review]
Lots of typing missed out. Didn't want to go through all of them so just this comment will suffice. 

I really like that you have moved the mouse functions out; it looks a lot cleaner! I'm not just debating whether or not it makes sense for them to be in the JigSaw class. `self.jigsaw.mouse_down`. This is all just semantics, but is JigSaw moving the mouse down, is the user, is it a tile that is moving down? I think this could be improved somewhat.

**Comment by EarthKiii** [comment]
These are events handlers, these are used when a click is captured by the JigSaw surface and if its released. This semantic makes sense to me.

**Comment by EarthKiii** [comment]
> Lots of typing missed out. Didn't want to go through all of them so just this comment will suffice.
> 
> I really like that you have moved the mouse functions out; it looks a lot cleaner! I'm not just debating whether or not it makes sense for them to be in the JigSaw class. `self.jigsaw.mouse_down`. This is all just semantics, but is JigSaw moving the mouse down, is the user, is it a tile that is moving down? I think this could be improved somewhat.

These are events handlers, these are used when a click is captured by the JigSaw surface and if its released. This semantic makes sense to me.

File: jigsaw-game/common/tessellation.py (hunk 1 above)
Comment (busterbeam) [reply jigsaw-game/common/tessellation.py:17]: We actually aren't using that many functions from numpy surprised that we didn't get something from flake8 or black. At this point you may be noticing that I don't care to follow the trend when I don't have to and the trend to use `np.` for all `numpy` is imports superfluous. We do import more so many functions from numpy that it does use the multiline import syntax I am willing to concede to the trend

File: jigsaw-game/client/jigsaw.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
+import client
```
Comment (LukeHankey) [inline jigsaw-game/client/jigsaw.py:1]:
```suggestion
from .tiles import Tile, JigSawTiles
```

File: jigsaw-game/client/jigsaw.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+from pygame import Rect, Surface
+
+TOP = BOTTOM = 1
+LEFT = RIGHT = 0
+
+
+class JigSaw(Surface):
+    """This is the Surface that the user interacts with the puzzle"""
+
+    def __init__(self, screen: Surface, size: tuple[int, int]):
```
Comment (LukeHankey) [inline jigsaw-game/client/jigsaw.py:11]:
```suggestion
    def __init__(self, screen: Surface, size: tuple[int, int]) -> None:
```

File: jigsaw-game/client/jigsaw.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+        """Set Jigsaw position"""
+        super(JigSaw, self).__init__(size)
+        self.screen = screen
+        self.size = size
+        self.set_bounds()
+
+    def set_bounds(self):
```
Comment (LukeHankey) [inline jigsaw-game/client/jigsaw.py:18]:
```suggestion
    def set_bounds(self) -> None:
```

File: jigsaw-game/client/jigsaw.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+        screen_bounds = self.screen.get_rect()
+        surface_bounds = self.get_rect()
+        x_offset = (screen_bounds.width - surface_bounds.width) // 2
+        y_offset = (screen_bounds.height - surface_bounds.height) // 2
+        self.rect = Rect(
+            x_offset, y_offset, self.size[0] + x_offset, self.size[1] + y_offset
+        )
+        self.center = (x_offset, y_offset)
+
+    def translate(self, position: tuple[int, int]):
```
Comment (LukeHankey) [inline jigsaw-game/client/jigsaw.py:29]:
```suggestion
    def translate(self, position: tuple[int, int]) -> None:
```
I think this and a few others are already covered in other PRs but not sure.

File: jigsaw-game/client/jigsaw.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+        )
+        return offset
+
+    def border_check_snap(self, side: int, tile: "client.Tile"):
+        """Checks if the tile can be snapped to the game border and snaps it if it can"""
+        if tile.rect.topleft[side] < tile.snapping_rect[side]:
+            return tile.snap("topleft", 0, side)
+        elif tile.rect.bottomright[side] > (self.size[side] - tile.snapping_rect[side]):
+            return tile.snap("bottomright", self.size[side], side)
+        return False
```
Comment (LukeHankey) [inline jigsaw-game/client/jigsaw.py:43]: This seems like it returns only None or False. If it's checking something, shouldn't this return either True or False?

File: jigsaw-game/client/jigsaw.py [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+
+    def translate(self, position: tuple[int, int]):
+        """Adjust screen coordinates to jigsaw surface coordinates"""
+        offset = (
+            position[0] - self.rect.topleft[0],
+            position[1] - self.rect.topleft[1],
+        )
+        return offset
+
+    def border_check_snap(self, side: int, tile: "client.Tile"):
```
Comment (LukeHankey) [inline jigsaw-game/client/jigsaw.py:37]:
```suggestion
    def border_check_snap(self, side: int, tile: Tile) -> bool | None:
```
Return type can be changed if the below comment has been addressed.

File: jigsaw-game/client/jigsaw.py [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+        """Checks if the tile can be snapped to the game border and snaps it if it can"""
+        if tile.rect.topleft[side] < tile.snapping_rect[side]:
+            return tile.snap("topleft", 0, side)
+        elif tile.rect.bottomright[side] > (self.size[side] - tile.snapping_rect[side]):
+            return tile.snap("bottomright", self.size[side], side)
+        return False
+
+    @staticmethod
+    def tile_check_snap(side: int, tile: "client.Tile", check_tile: "client.Tile"):
```
Comment (LukeHankey) [inline jigsaw-game/client/jigsaw.py:46]: What's the use here of making it a staticmethod? These aren't very useful and rarely used.
```suggestion
    def tile_check_snap(side: int, tile: Tile, check_tile: Tile) -> bool | None:
```
Same as above with the retuning of True/False.

File: jigsaw-game/client/jigsaw.py [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+        bottomright_snap = abs(
+            check_tile.rect.topleft[side] - tile.rect.bottomright[side]
+        )
+        if topleft_snap < bottomright_snap and topleft_snap <= tile.snapping_rect[side]:
+            return tile.snap("topleft", check_tile.rect.bottomright[side], side)
+        elif bottomright_snap <= tile.snapping_rect[side]:
+            return tile.snap("bottomright", check_tile.rect.topleft[side], side)
+        return False
+
+    def mouse_down(self, tiles: "client.JigSawTiles", mouse_position: tuple[int, int]):
```
Comment (LukeHankey) [inline jigsaw-game/client/jigsaw.py:58]:
```suggestion
    def mouse_down(self, tiles: JigSawTiles, mouse_position: tuple[int, int]) -> JigSawTiles:
```

File: jigsaw-game/client/jigsaw.py [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+        for tile in tiles.sprites()[::-1]:
+            if tile.rect.collidepoint(self.translate(mouse_position)):
+                # reorder to the top before drag
+                tiles.remove(tile)
+                tiles.add(tile)  # reorder to top
+                tile.activate(mouse_position)
+                break
+        return tiles
+
+    def mouse_up(self, tiles: "client.JigSawTiles", mouse_position: tuple[int, int]):
```
Comment (LukeHankey) [inline jigsaw-game/client/jigsaw.py:70]:
```suggestion
    def mouse_up(self, tiles: JigSawTiles, mouse_position: tuple[int, int]) -> None:
```

File: jigsaw-game/client/tiles.py [hunk 12]
```diff
@@ -0,0 +1,67 @@
+from pygame import image, sprite
+
+
+class JigSawTiles(sprite.Group):
+    """Inheritance for the addition of Group methods concerning active tiles"""
+
+    def get_active(self):
```
Comment (LukeHankey) [inline jigsaw-game/client/tiles.py:7]:
```suggestion
    def get_active(self) -> Tile | False:
```
I don't think this should return `False`. `None` would be better.

File: jigsaw-game/client/tiles.py [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+        """Return the active tile"""
+        for tile in self.sprites()[::-1]:
+            if tile.active:
+                return tile
+        return False
+
+    def get_inactives(self):
```
Comment (LukeHankey) [inline jigsaw-game/client/tiles.py:14]:
```suggestion
    def get_inactives(self) -> list[Tile]:
```

File: jigsaw-game/client/jigsaw.py [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+        )
+        self.center = (x_offset, y_offset)
+
+    def translate(self, position: tuple[int, int]):
+        """Adjust screen coordinates to jigsaw surface coordinates"""
+        offset = (
+            position[0] - self.rect.topleft[0],
+            position[1] - self.rect.topleft[1],
+        )
+        return offset
```
Comment (LukeHankey) [inline jigsaw-game/client/jigsaw.py:35]:
```suggestion
        return position[0] - self.rect.topleft[0], position[1] - self.rect.topleft[1]
```

File: jigsaw-game/client/tiles.py [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+        """Return the inactive tiles"""
+        return [tile for tile in self.sprites() if not tile.active]
+
+
+class Tile(sprite.Sprite):
```
Comment (LukeHankey) [inline jigsaw-game/client/tiles.py:19]: This should already be typed in another PR.

File: jigsaw-game/common/tessellation.py (hunk 1 above)
Comment (LukeHankey) [reply jigsaw-game/common/tessellation.py:17]: Why would you not want to use common trends? They are established for a reason; in particular to avoid name clashes. If you think this is something that won't affect the project from the reviewers, then it should be fine to keep. I would advise though to stick to these "trends" as often as possible.

File: jigsaw-game/common/tessellation.py [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+            hstack([x_edge, y_edge_top]),
+            hstack([x_edge, y_edge_bottom]),
+            hstack([x_edge_left, y_edge]),
+            hstack([x_edge_right, y_edge]),
+        ]
+    )
+    return vstack([random.rand(number_of_points, 2), edge_points])
+
+
+def square_grid(width: int, height: int, num_of_tiles: int) -> array:
```
Comment (LukeHankey) [inline jigsaw-game/common/tessellation.py:46]: Not at my own laptop atm but is this return type valid?

**Review by LukeHankey — CHANGES_REQUESTED** [review]

File: jigsaw-game/common/tessellation.py (hunk 1 above)
Comment (busterbeam) [reply jigsaw-game/common/tessellation.py:17]: I am fully aware of name clashing, but I hardly see that happening in this project, at best if we import `math.sqrt` alongside `numpy.sqrt` or the admittedly more real issues we could get from `scipy`, but we should be decisive. to use the correct functions from both `numpy` and `scipy`  and the types/parameters are cross compatible so no issues there I think.  But I do appreciate you bringing it up

**Comment by busterbeam** [comment]
> Lots of typing missed out. Didn't want to go through all of them so just this comment will suffice.
> 
> I really like that you have moved the mouse functions out; it looks a lot cleaner! I'm not just debating whether or not it makes sense for them to be in the JigSaw class. `self.jigsaw.mouse_down`. This is all just semantics, but is JigSaw moving the mouse down, is the user, is it a tile that is moving down? I think this could be improved somewhat.

I back @EarthKiii here as he is the one working this mainly, I may of changed it but if he likes the change he gets final say.

File: jigsaw-game/client/jigsaw.py (hunk 6 above)
Comment (busterbeam) [reply jigsaw-game/client/jigsaw.py:29]: No this is supposed to return tuple of int's

Comment (LukeHankey) [reply jigsaw-game/client/jigsaw.py:29]: You are correct. An oversight on my part.

File: jigsaw-game/client/jigsaw.py (hunk 7 above)
Comment (busterbeam) [reply jigsaw-game/client/jigsaw.py:43]: sorry, this is something that should be fixed in `tile.snap` return `True` if snap is successful. the snap should be successful if those "checks" are done before hand. Hopefully @EarthKiii is on this

**Comment by EarthKiii** [comment]
i feel like 50% is too high for the snapping window, maybe half ?

File: requirements.txt (hunk 2 above)
Comment (busterbeam) [reply requirements.txt:5]: I find `numpy` gets installed with `scipy` by default

File: jigsaw-game/client/tiles.py (hunk 15 above)
Comment (busterbeam) [reply jigsaw-game/client/tiles.py:4]: can objects be typed? or will this be referencing something else?

File: jigsaw-game/common/tessellation.py (hunk 16 above)
Comment (busterbeam) [reply jigsaw-game/common/tessellation.py:13]: valid return type

**Comment by busterbeam** [comment]
> i feel like 50% is too high for the snapping window, maybe half ?

50% is half. if the tile moves 50% down the tile space it covers the next tile by about half.  is it that you want tiles to still partially overlap?

**Comment by EarthKiii** [comment]
> > i feel like 50% is too high for the snapping window, maybe half ?
> 
> 50% is half. if the tile moves 50% down the tile space it covers the next tile by about half.  is it that you want tiles to still partially overlap?
> 
> 50% prevents partial overlaps. 
> 
> 

Well since there isn't any space to temporarily put the tiles while moving them yes, that overlap is important

**Review by LukeHankey — APPROVED** [review]
